lib/Crypto/doc/sailfish-crypto-overview.qdoc

```diff
@@ -0,0 +1,218 @@
+\li Securely generate and store a cryptographic key based upon some template
+\li Access a securely stored cryptographic key
+\li Delete a securely stored cryptographic key
+\li Sign arbitrary data using a cryptographic key
+\li Verify a signature using a cryptographic key or certificate
+\li Encrypt arbitrary data using a cryptographic key
+\li Decrypt encrypted data using a cryptographic key
+\endlist
+
+\section3 Usage Examples
+
+The examples directory in the source repository contains a variety of examples
+of usage of the Sailfish OS Crypto Library as well as the Sailfish OS Secrets
+Library.  Please see those for complete, working examples.
+
+Some snippets showing commonly-required functionality are included below.
+
+\section4 Generating a securely-stored symmetric cipher key
+
+This snippet shows how to generate a symmetric cipher key which will be stored
+securely by the Sailfish OS Secrets and Crypto Framework. It assumes that a
+secure storage collection has previously been created, in which to store the
+key. Please see the Sailfish OS Secrets Library documentation for information
+about how to create such a collection.
+
+First, the client defines the algorithm, supported block modes, supported
+padding schemes, and supported operations in a template key.  That template key
+is given an identifier which includes the name of the key and the name of the
+secure storage collection in which it should be stored, and is then passed as a
+parameter when invoking the \tt{generateStoredKey()} method, which results in an
+IPC call to the secure Sailfish OS Secrets and Crypto Framework system service.
+
+The Sailfish OS Secrets and Crypto Framework system service will then delegate
+the operation to the specified crypto plugin, which in turn will generate a
+full symmetric key based upon the given template, and store it securely.  A
+reference key (that is, a key containing a valid identifier and metadata, but
+no secret key data) will be returned to the client application.
+
+Note that these operations are all asynchronous, however in the snippet we
+force the operation to be synchronous by calling the \tt{waitForFinished()}
+method on the reply object.  In practice, the client application should instead
+use a watcher object which will notify when the operation is complete.
+
+\code
+// Set the key template metadata.
+Sailfish::Crypto::Key keyTemplate, symmetricKeyReference;
+keyTemplate.setAlgorithm(Sailfish::Crypto::Key::Aes256);
+keyTemplate.setOrigin(Sailfish::Crypto::Key::OriginDevice);
+keyTemplate.setBlockModes(Sailfish::Crypto::Key::BlockModeCBC);
+keyTemplate.setEncryptionPaddings(Sailfish::Crypto::Key::EncryptionPaddingNone);
+keyTemplate.setSignaturePaddings(Sailfish::Crypto::Key::SignaturePaddingNone);
+keyTemplate.setDigests(Sailfish::Crypto::Key::DigestSha256);
+keyTemplate.setOperations(Sailfish::Crypto::Key::Encrypt
+                         |Sailfish::Crypto::Key::Decrypt);
```

Comment:
While repeating the namespaces is good for teaching the namespace structure it makes the example seem unnecessarily complex. I wish C++ supported Key key(algorithm=Aes256, origin=OriginDevice, blockmodes=BlockModeCBC, ..)  -style syntax, where the order of parameters didn't matter.

Reply:
@jpetrell It can be solved by putting `using namespace Sailfish::Crypto;` into the scope. Then it will be just `Key::Aes256`, for example. The other way is to make a typedef/using: `using Key = Sailfish::Crypto::Key;` and then just `Key::Aes256`.

Also, the use of knowing the exact fully qualified path is under doubts. Boost has namespaces too, but everyone and even they themselves use aliases to their types, like `using po = boost::program_options;`. People will know the path but there are doubts anyone would use the full path in each line (saves key presses and time at the very least.

Also, by having these unspoken rules where namespace names are in UpperCamelCase it is impossible to know what is the type and what is the namespace for the developer: that is why everyone uses lower camel case or lower case with underscores for the namespaces. Qt uses UpperCamelCase only for the types, not namespaces. I have tried to find a rule where you could possibly see it but there is nothing like that. And I don't remember that it was something like that there.

Reply:
Qt does use camel-cased namespaces, e.g. "QPatternist", "QtAndroid", "QtMobility", "Simulator", "QDBus", "QMdi", etc etc.

Yes, the developer can have "using namespace" however that can cause issues when using both the Crypto and the Secrets API, since some type names are shared (which is a major reason to use namespaces, avoiding polluting the global namespace).  Yes, using full symbol naming makes the examples more verbose BUT it also makes them more precise.  We should assume that our client developers are intelligent and able to figure out how to improve the readability of their own code in the manner they see fit.

Reply:
So I’d use short type alias.